process/sync/metaForkDetector.go

```diff
@@ -69,31 +74,52 @@ func (mfd *metaForkDetector) AddHeader(
 		return err
 	}
 
-	mfd.activateForcedForkOnConsensusStuckIfNeeded(header, state)
+	if header.GetNonce() > mfd.highestNonceReceived() {
+		log.Debug("forkDetector.AddHeader.setHighestNonceReceived",
+			"nonce", header.GetNonce(),
+			"highest nonce received", mfd.highestNonceReceived())
+		mfd.setHighestNonceReceived(header.GetNonce())
+	}
 
-	isHeaderReceivedTooLate := mfd.isHeaderReceivedTooLate(header, state, process.MetaBlockFinality)
-	if isHeaderReceivedTooLate {
-		state = process.BHReceivedTooLate
+	if state == process.BHProposed {
+		mfd.activateForcedForkOnConsensusStuckIfNeeded(header, state)
```

Comment:
This part has been refactored in my last commit. There was an old bug

Reply:
the bug should be explained, it is hard to understand why these changes are correct.

Reply:
Actually this was one of the problem here. We should called isHeaderReceivedTooLate method only for received headers (BHReceived). Because we did this even for the proposed headers received BHProposed (which are used now only to trigger forced fork if the consensus is stuck), many of them would have its state changed form BHProposed in BHReceivedTooLate and the trigger has been avoided. The second problem, which was more important, was that if the consensus was stuck, and in the round equal with RoundModulusTrigger no block was propossed the forced fork would not have been forced because nobody would set this trigger. So now the refactored model will work much simple. The check and set would be done in CheckFork which is called in the beginning of each round no matter what. So now we will have the check/set/action in the same round (RoundModulusTrigger) by all nodes no matter what. In the old version the check and set for a forced fork was done in AddHeader when round was equal with RoundModulusTrigger (and only if the node would received a proposed block. Because of this condition it would been possible that not all nodes to set this action in the case they would not been received this message on consensus topic due the network latency or network split). After this check/set the real action would be done only in RoundModulusTrigger+1 only if the set has been done in last round. This would caused some edge cases when one proposed block would be accepted, after many rounds without successfully consensus, exactly in the RoundModulusTrigger, after all node would set the forced fork for the next round.